akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala

```diff
@@ -54,212 +60,86 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
       override def updateDefaultWindow(newDefaultWindow: Int): Unit = {
-        val delta = newDefaultWindow - currentInitialWindow
+        val delta = newDefaultWindow - _currentInitialWindow
+        _currentInitialWindow = newDefaultWindow
 
-        currentInitialWindow = newDefaultWindow
-        outStreams.values.foreach(i => updateWindowFor(i.streamId, delta))
+        distributeWindowDeltaToAllStreams(delta)
       }
       override def updatePriority(info: PriorityFrame): Unit = prioritizer.updatePriority(info)
 
-      private def streamFor(streamId: Int): OutStream = outStreams.get(streamId) match {
-        case None =>
-          val newOne = new OutStream(streamId, None, currentInitialWindow)
-          outStreams += streamId -> newOne
-          newOne
-        case Some(old) => old
-      }
-      private def windowLeftFor(streamId: Int): Int = streamFor(streamId).outboundWindowLeft
-      private def updateWindowFor(streamId: Int, increment: Int): Unit = {
-        val info = streamFor(streamId)
-        info.outboundWindowLeft += increment
-        if (info.canSend) enqueueOutStream(info)
-      }
+      def enqueueOutStream(streamId: Int): Unit = updateState(_.enqueueOutStream(streamId))
+      def closeStream(streamId: Int): Unit = updateState(_.closeStream(streamId))
 
-      def enqueueOutStream(outStream: OutStream): Unit = state.enqueueOutStream(outStream)
-      def closeStream(outStream: OutStream): Unit = state.closeStream(outStream)
+      def onPull(): Unit = updateState(_.onPull())
 
-      override def shutdown(): Unit = outStreams.values.foreach(_.cancelStream())
+      private var _state: MultiplexerState = Idle
 
-      var state: MultiplexerState = Idle
-      def onPull(): Unit = state.onPull()
-      private def become(nextState: MultiplexerState): Unit = {
-        if (nextState.name != state.name) recordStateChange(state.name, nextState.name)
+      private def updateState(transition: MultiplexerState => MultiplexerState): Unit = {
+        val oldState = _state
+        val newState = transition(_state)
+        _state = newState
 
-        state = nextState
+        if (newState.name != oldState.name) recordStateChange(oldState.name, newState.name)
       }
 
       private[http2] sealed trait MultiplexerState extends Product {
         def name: String = productPrefix
 
-        def onPull(): Unit
-        def pushControlFrame(frame: FrameEvent): Unit
-        def connectionWindowAvailable(): Unit
-        def enqueueOutStream(outStream: OutStream): Unit
-        def closeStream(outStream: OutStream): Unit
+        def onPull(): MultiplexerState
+        def pushControlFrame(frame: FrameEvent): MultiplexerState
+        def connectionWindowAvailable(): MultiplexerState
+        def enqueueOutStream(streamId: Int): MultiplexerState
+        def closeStream(streamId: Int): MultiplexerState
+
+        protected def sendDataFrame(streamId: Int, sendableOutstreams: immutable.Set[Int]): MultiplexerState = {
```

Comment:
The other is the greatly simplified handling of sending outgoing data and the interface with `OutStream` / stream handling. The full interface is now:

 * stream handling can call `enqueueOutStream` if a stream wants to send a data frame
 * if there's connection window and the network pulls and there are no control frames to send, the multiplexer state machine calls `sendDataFrame` here. This calls `pullNextFrame` to get the next frame and more info about what the state of the stream is afterwards.
 * stream handling can call `closeStream` to remove a potentially enqueued stream